keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -0,0 +1,462 @@
+   - for fields with `x-kubernetes-list-type: set` the order is ignored.
+   
+   If that comparison fails, the request is rejected as 400 "Bad Request".
+   
+We create another KEP to define custom normalization steps for CRs done during 
+deserialization from etcd and when receiving a request (just after pruning and defaulting).
+
+### OpenAPI extension `x-kubernetes-immutable`
+
+The `x-kubernetes-immutable` vendor extension is set in the spec of fields,
+arrays and objects. It recursively applies to the whole subtree:
+
+```yaml
+properties:
+  foo:
+    type: array
+    items:
+      x-kubernetes-immutable: true
+      type: string
+```
+
+means that the `foo` array itself is mutable, but the items are not. This means that
+items can be deleted and added, but at each existing index the string values are
+immutable.
+
+```yaml
+properties:
+  foo:
+    type: object
+    x-kubernetes-immutable: true
+```
+
+means that the whole `.foo` object is immutable. It cannot be removed or set.
+
+If one wants to make only field immutable, but allow to delete an existing object
+or set it if unset, the following is used:
+
+```yaml
+properties:
+  foo:
+    type: object
+    properties:
+      "x":
+        type: string
+        x-kubernetes-immutable: true
+      "y":
+        type: string
+        x-kubernetes-immutable: true
+```
+
+for all fields `x` and `y` of `foo`.
+
+```yaml
+properties:
+  foo:
+    type: object
+    x-kubernetes-immutable: true
+    properties:
+      bar:
+        type: object
+        x-kubernetes-immutable: true
+```
+
+can be simplified to 
+
+```yaml
+properties:
+  foo:
+    type: object
+    x-kubernetes-immutable: true
+    properties:
+      bar:
+        type: object
+```
+
+At the root-level and inside `.metadata` of an object, `x-kubernetes-immutable` is forbidden.
+
+### OpenAPI extension `x-kubernetes-immutable-keys`
+
+The topology of arrays is specified via `x-kubernetes-list-type` and `x-kubernetes-list-map-keys`. We currently support the `atomic`, `set` and `map` types. These interact with immutability in the following way:
+
+- **Non-recursive** immutability: For an array or a map, one can not remove or add new items/values, but existing items/values can be modified. 
+
+  - For maps this means that the the keys stay the same, but values might change:
+    ```yaml
+    properties:
+      someMap:
+        type: object
+        x-kubernetes-immutable-keys: true # allowed with additionalProperties
+        additionalProperties:
+          type: string
+    ```
+    with `x-kubernetes-immutable-keys` allowed due to `additionalProperties` and mutually
+    exclusive with `x-kubernetes-immutable` on the same node.
+    
+  - For arrays of list-type `map` (associative lists) we get the same behaviour, but enforce that the map-keys
+    are immutable as well (via CRD validation):
```

Comment:
> but enforce that the map-keys are immutable as well

isn't this true for the case above as well?

Reply:
No, in the upper case only the keys are immutable, i.e. the set of keys. But values can change.